packages/notifications/push/amplify_push_notifications_pinpoint/example/ios/unit_tests/AmplifyPushNotificationsPluginTest.swift

```diff
@@ -0,0 +1,530 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import XCTest
+import Flutter
+
+@testable import amplify_push_notifications
+
+let fakePointer = AutoreleasingUnsafeMutablePointer<FlutterError?>(bitPattern: 1)!
+let expectedNotification = [
+    "aps": [
+        "alert": ["title": "I'm a PN"],
+        "content-available": 1
+    ] as [String : Any]
+]
+let expectedDeviceToken = "abcd1234ef5678";
+let expectedDeviceTokenData = revertTokenOperation(expectedDeviceToken)
+let expectedErrorLocalizedDescription = NSLocalizedString("register token failed", comment: "mockError")
+var fakeIconBadgeNumber = 0
+var fakeUIApplicationState = UIApplication.State.active
+var hasCalledRegisterForRemoteNotifications = false
+var hasOnNotificationReceivedInBackgroundBeenCalled = false
+var hasNotificationOpenedHandlerBeenCalled = false
+var hasForegroundMessageReceivedHandlerBeenCalled = false
+var hasTokenReceivedHandlerBeenCalled = false
+var hasTokenReceivedErrorHandlerBeenCalled = false
+
+final class AmplifyPushNotificationsPluginTest: XCTestCase {
+    var originalApplicationStateGetter: Method?
+    var mockedApplicationStateGetter: Method?
+    var pluginInstance: AmplifyPushNotificationsPlugin?
+
+    override func setUp() {
+        // Put setup code here. This method is called before the invocation of each test method in the class.
+        let registrar = MockFlutterPluginRegistrar()
+        AmplifyPushNotificationsPlugin.register(with: registrar);
+
+        let pluginInstance = registrar.instance as? AmplifyPushNotificationsPlugin
+        self.pluginInstance = pluginInstance
+
+        originalApplicationStateGetter = class_getInstanceMethod(
+            UIApplication.self,
+            #selector(getter: UIApplication.applicationState)
+        )
+        mockedApplicationStateGetter = class_getInstanceMethod(
+            MockUIApplicationStateProvider.self,
+            #selector(getter: MockUIApplicationStateProvider.applicationState)
+        )
+        method_exchangeImplementations(originalApplicationStateGetter!, mockedApplicationStateGetter!)
```

Comment:
Thoughts on using a proper Swift mocking framework?

Reply:
`UIApplication` is restricted to have one single instance in an App life cycle, mocking it is easy, but cannot instantiate the mock running tests. So I had to use method swizzle to swap out the method impl. on the same instance.